rust/src/tests/rename.rs

```diff
@@ -0,0 +1,275 @@
+
+    assert_ne!(symlink_stat.to_invariant(), sym_target_stat.to_invariant());
+    assert_eq!(sym_target_stat.to_invariant(), target_stat.to_invariant());
+
+    let sym_new_path = ctx.base_path().join("sym_new_path");
+    rename(&symlink_old_path, &sym_new_path).unwrap();
+
+    let sym_target_stat = stat(&sym_new_path).unwrap();
+    assert_eq!(target_stat, sym_target_stat);
+    assert_eq!(lstat(&symlink_old_path).unwrap_err(), Errno::ENOENT);
+
+    let sym_new_stat = lstat(&sym_new_path).unwrap();
+    assert_eq!(symlink_stat.to_invariant(), sym_new_stat.to_invariant());
+}
+
+crate::test_case! {
+    /// rename should not update ctime if it fails
+    // rename/00.t
+    unchanged_ctime_failed, serialized, root => [Regular, Dir, Fifo, Block, Char, Socket, Symlink(None)]
+}
+fn unchanged_ctime_failed(ctx: &mut SerializedTestContext, ft: FileType) {
+    let file = ctx.new_file(ft).mode(0o600).create().unwrap();
+    let other_path = ctx.gen_path();
+    let user = ctx.get_new_user();
+    ctx.as_user(&user, None, || {
+        assert_symlink_ctime_unchanged(ctx, &file, || {
+            assert!(rename(&file, &other_path).is_err());
+        })
+    });
+}
+
+crate::test_case! {
+    /// write access to subdirectory is required to move it to another directory
+    // rename/21.t
+    write_access_required_subdir, serialized, root
+}
+fn write_access_required_subdir(ctx: &mut SerializedTestContext) {
+    let dir = ctx.new_file(FileType::Dir).mode(0o777).create().unwrap();
+    let subdir = ctx
+        .new_file(FileType::Dir)
+        .name(dir.join("subdir"))
+        .mode(0o700)
+        .create()
+        .unwrap();
+    let another_subdir_path = dir.join("another_subdir_path");
+
+    let new_dir = ctx.new_file(FileType::Dir).mode(0o777).create().unwrap();
+    let new_dir_subpath = new_dir.join("subpath");
+
+    let user = ctx.get_new_user();
+    ctx.as_user(&user, None, || {
+        // Check that write permission on containing directory is enough
+        // to rename subdirectory. If we rename directory write access
+        // to this directory may also be required.
+        assert!(matches!(
+            rename(&subdir, &another_subdir_path),
+            Ok(_) | Err(Errno::EACCES)
+        ));
+
+        assert!(matches!(
+            rename(&another_subdir_path, &subdir),
+            Ok(_) | Err(Errno::EACCES)
+        ));
+
+        // Check that write permission on containing directory is not enough
+        // to move subdirectory from that directory.
+        // Actually POSIX says that write access to `dir` and `new_dir` may be enough
+        // to move it.
+        assert!(matches!(
+            rename(&subdir, &new_dir_subpath),
+            Ok(_) | Err(Errno::EACCES)
+        ));
+    });
+
+    // Check that write permission on containing directory (${n2}) is enough
+    // to move file (${n0}) from that directory.
+    let file = ctx
+        .new_file(FileType::Regular)
+        .name(dir.join("file"))
```

Comment:
I think you should explicitly set the file's mode here, because it's important that `nobody` cannot write to it.